digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MaintenanceService.scala

```diff
@@ -0,0 +1,267 @@
+
+    if (newAssets.nonEmpty) {
+      logger.info("Transferred %d assets in %d ms ".format(newAssets.length, System.currentTimeMillis - timing))
+    }
+    val groupedAssets = (existingAssets.filterNot(a => expiredPavingAssetIds.contains(a.id) || newAssets.exists(_.linkId == a.linkId)) ++ newAssets).groupBy(_.linkId)
+    val (filledTopology, changeSet) = NumericalLimitFiller.fillTopology(roadLinks, groupedAssets, maintenanceRoadAssetTypeId)
+
+    val expiredAssetIds = existingAssets.filter(asset => removedLinkIds.contains(asset.linkId)).map(_.id).toSet ++
+      changeSet.expiredAssetIds ++ expiredPavingAssetIds
+
+    val mValueAdjustments = newAndUpdatedPavingAssets.filter(_.id != 0).map( a =>
+      MValueAdjustment(a.id, a.linkId, a.startMeasure, a.endMeasure)
+    )
+    eventBus.publish("linearAssets:update", changeSet.copy(expiredAssetIds = expiredAssetIds.filterNot(_ == 0L),
+      adjustedMValues = changeSet.adjustedMValues ++ mValueAdjustments))
+
+    //Remove the asset ids ajusted in the "linearAssets:update" otherwise if the "linearAssets:saveProjectedLinearAssets" is executed after the "linearAssets:update"
+    //it will update the mValues to the previous ones
+    eventBus.publish("linearAssets:saveProjectedLinearAssets", newAssets.filterNot(a => changeSet.adjustedMValues.exists(_.assetId == a.id)))
+
+    filledTopology
+  }
+
+  protected def getLinkSource(roadlink: Option[RoadLink]): Option[Int] = {
+    roadlink match {
+      case Some(result) =>
+        Some(roadlink.get.linkSource.value)
+      case _ => None
+    }
+  }
+
+  /**
+    * Returns Maintenance assets by asset type and asset ids.
+    */
+  override def getPersistedAssetsByIds(typeId: Int, ids: Set[Long]): Seq[PersistedLinearAsset] = {
+    withDynTransaction {
+      maintenanceDAO.fetchMaintenancesByIds(maintenanceRoadAssetTypeId, ids)
+    }
+  }
+
+  def getPersistedAssetsByLinkIds(linkIds: Seq[Long]): Seq[PersistedLinearAsset] = {
+    withDynTransaction {
+      maintenanceDAO.fetchMaintenancesByLinkIds(maintenanceRoadAssetTypeId, linkIds, includeExpire = false)
+    }
+  }
+
+  override def getUncheckedLinearAssets(areas: Option[Set[Int]]): Map[String, Map[String ,List[Long]]] ={
+    val unchecked = withDynTransaction {
+      maintenanceDAO.getUncheckedMaintenanceRoad(areas)
+
+    }.groupBy(_._2).mapValues(x => x.map(_._1))
+    Map("Unchecked" -> unchecked )
+  }
+
+  /**
+    * Saves linear asset when linear asset is split to two parts in UI (scissors icon). Used by Digiroad2Api /linearassets/:id POST endpoint.
+    */
+  override def split(id: Long, splitMeasure: Double, existingValue: Option[Value], createdValue: Option[Value], username: String, municipalityValidation: (Int) => Unit): Seq[Long] = {
+    withDynTransaction {
+      val linearAsset = dao.fetchLinearAssetsByIds(Set(id), LinearAssetTypes.numericValuePropertyId).head
```

Comment:
Since we are inside of `MaintenanceService`, shouldn't this be ` maintenanceDAO.fetchMaintenancesByIds`

Reply:
This is a good point, we have to refactor the linear asset to have a extend for each type of linear assets. Let's have it in  account in the next one.